raft/node.go

```diff
@@ -114,11 +114,25 @@ type Node interface {
 	ProposeConfChange(ctx context.Context, cc pb.ConfChange) error
 	// Step advances the state machine using the given message. ctx.Err() will be returned, if any.
 	Step(ctx context.Context, msg pb.Message) error
+
 	// Ready returns a channel that returns the current point-in-time state.
-	// Users of the Node must call Advance after applying the state returned by Ready.
+	// Users of the Node must call Advance after retrieving the state returned by Ready.
+	//
+	// NOTE: No committed entries from the next Ready may be applied until all committed entries
+	// and snapshots from the previous one have finished.
 	Ready() <-chan Ready
-	// Advance notifies the Node that the application has applied and saved progress up to the last Ready.
+	// Advance notifies the Node that the application has saved progress up to the last Ready.
 	// It prepares the node to return the next available Ready.
+	//
+	// In normal case, application should call Advance after it applies the entries in last Ready.
+	// When the last Ready contains a snapshot, the application might take a long time to apply
+	// the snapshot data. To continue receiving Ready without blocking raft progress, it is safe(*)
+	// to call Advance before finish applying the last ready. But the applied index in your application
+	// will be behind the applied index that tracks by Node.
+	//
+	// (*) NOTE: If a follower changes its state to candidate, it MUST wait for finishing all previous
```

Comment:
> It feels weird that there wouldn't be a way where I can simply apply changes to my state machine asynchronously, without jumping through this (seemingly tricky) hoop of checking whether my node is a candidate.

You can apply log entries asynchronously. The only problem is that we need to ensure there is no more than one unapplied configuration change in the log of leader. So after the application receives the state change from follower to candidate (even before it can become a leader), it should block and wait for all pending configurations to be applied if there is any. I think this is very easy to achieve.